### PR TITLE
Fixed modelc exception handling for Win32

### LINKED ISSUE #7294: Collection becomes unresponsive when attaching an animation set to a model component using GLB files
### Expected behaviour
Should be able to attach an `.animationset` file to a model component consisting of GLB files with no errors, allowing you to play those animations.

### Actual behaviour
Attaching an animation set to the model component throws a Clojure error and causes the affected collection file to become unresponsive. Building the game also throws an error.

You can reverse the change by editing the text of the collection file to remove the animation set and restarting Defold.

### Steps to reproduce
1. Download the [attached Defold project](https://github.com/defold/defold/files/10462965/animsetbug.zip). It's an empty project with only the affected files.
2. Open `main.collection`.
3. Click on the model component, and attach `ybot.animationset` to the Animations property.

<hr/>

<table>
<tr><td>Build time</td><td>2023-01-19T21:57:02.089212</td></tr>
<tr><td>Defold channel</td><td>editor-alpha</td></tr>
<tr><td>Defold editor sha</td><td>26e4e30870571b201415a84178387305d91d630b</td></tr>
<tr><td>Defold engine sha</td><td>8f96e450ddfb006a99aa134fdd373cace3760571</td></tr>
<tr><td>Defold version</td><td>1.4.1</td></tr>
<tr><td>Error</td><td><a href='https://sentry.io/organizations/defold/issues/?query=id%3A"c4d938f09b0f48f388708173ce4f68a5"'>c4d938f09b0f48f388708173ce4f68a5</a></td></tr>
<tr><td>GPU</td><td>Apple M1</td></tr>
<tr><td>GPU Driver</td><td>2.1 Metal - 83</td></tr>
<tr><td>Java version</td><td>11.0.15+10-LTS</td></tr>
<tr><td>OS arch</td><td>x86_64</td></tr>
<tr><td>OS name</td><td>Mac OS X</td></tr>
<tr><td>OS version</td><td>13.1</td></tr>
<table>

### PR DESCRIPTION
**Pull request by JCash**
We fixed a use-after-free issue that was noticed on Windows platforms.

Fixes #7294

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

### CODE REVIEW

File: build_tools/sdk.py [hunk 1]
```diff
@@ -219,10 +219,10 @@ def get_windows_local_sdk_info(platform):
     if windows_info is not None:
         return windows_info
 
-    vswhere_path = '%s/../../scripts/windows/vswhere2/vswhere2.exe' % os.environ['DYNAMO_HOME']
+    vswhere_path = '%s/../../scripts/windows/vswhere2/vswhere2.exe' % os.environ.get('DYNAMO_HOME', '.')
     if not os.path.exists(vswhere_path):
         vswhere_path = './scripts/windows/vswhere2/vswhere2.exe'
-        vswhere_path = path.normpath(vswhere_path)
+        vswhere_path = os.path.normpath(vswhere_path)
```
Comment (JCash) [inline build_tools/sdk.py:225]: Makes the Visual Studio detection nicer by not throwing an exception to the user

File: engine/modelc/src/jni_util.cpp [hunk 2]
```diff
@@ -75,7 +75,7 @@ LONG WINAPI ExceptionHandler(EXCEPTION_POINTERS* ptr)
 {
     g_ExceptionPointers = ptr;
     g_UserCallback(0xDEAD, g_UserContext);
-    return EXCEPTION_CONTINUE_SEARCH;
+    return EXCEPTION_EXECUTE_HANDLER;
```
Comment (JCash) [inline engine/modelc/src/jni_util.cpp:78]: It means we're handling it, no need to look further.

File: engine/modelc/src/jni_util.h [hunk 3]
```diff
@@ -28,6 +34,26 @@ namespace dmJNI
...
+#if defined(_WIN32)
+    LONG WINAPI ExceptionHandler(EXCEPTION_POINTERS* ptr);
+
+    #define DM_JNI_GUARD_SCOPE_BEGIN() \
+        __try { \
+
+    #define DM_JNI_GUARD_SCOPE_END(...) \
+        } __except (dmJNI::ExceptionHandler(GetExceptionInformation())) { \
+            __VA_ARGS__ \
+        }
```
Comment (JCash) [inline engine/modelc/src/jni_util.h:47]: The MSVC intrinsics `__try`/`__except` managed to catch this access violation, whereas the previous code didn't.

File: engine/modelc/src/modelimporter_gltf.cpp [hunk 4]
```diff
@@ -1047,7 +1057,7 @@ Scene* LoadGltfFromBuffer(Options* importeroptions, void* mem, uint32_t file_siz
 
     for (cgltf_size i = 0; i < data->buffers_count; ++i)
     {
-        scene->m_Buffers[i].m_Uri = data->buffers[i].uri;
+        scene->m_Buffers[i].m_Uri = CreateObjectName(data->buffers, "buffer", i);
```
Comment (JCash) [inline engine/modelc/src/modelimporter_gltf.cpp:1060]: Bugfix, if null, we need to create a generated name for this buffer.

File: engine/modelc/src/modelimporter_jni.cpp [hunk 5]
```diff
@@ -1013,6 +1006,20 @@ JNIEXPORT jobject JNICALL Java_ModelImporter_LoadFromBufferInternal(JNIEnv* env,
...
+
+JNIEXPORT jobject JNICALL Java_ModelImporter_LoadFromBufferInternal(JNIEnv* env, jclass cls, jstring _path, jbyteArray array, jobject data_resolver)
+{
+    dmLogDebug("Java_ModelImporter_LoadFromBufferInternal: env = %p\n", env);
+    dmJNI::SignalContextScope env_scope(env);
+
+    jobject jscene;
+    DM_JNI_GUARD_SCOPE_BEGIN();
+        jscene = LoadFromBufferInternal(env, cls, _path, array, data_resolver);
+    DM_JNI_GUARD_SCOPE_END(return 0;);
```
Comment (JCash) [inline engine/modelc/src/modelimporter_jni.cpp:1022]: Moved the code to an internal function, in order to make it easier  to wrap the code with guards (and also more readable).

**Review by Jhonnyg — DISMISSED** [review]
👍

**Comment by JCash** [comment]
Fixes https://github.com/defold/defold/issues/7294

**Pull request merged**